pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,662 @@
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+		},
+		/// Event emitted when an order is fulfilled.
+		/// Can be for either partial or total fulfillment.
+		/// Contains amount fulfilled, and whether fulfillment was partial or
+		/// full.
+		OrderFulfillment {
+			order_id: T::Hash,
+			placing_account: T::AccountId,
+			fulfilling_account: T::AccountId,
+			partial_fulfillment: bool,
+			fulfillment_amount: T::ForeignCurrencyBalance,
+			currency_in: T::AssetCurrencyId,
+			currency_out: T::AssetCurrencyId,
+			sell_price_limit: T::ForeignCurrencyBalance,
+		},
+	}
+
+	#[pallet::error]
+	#[derive(PartialEq)]
+	pub enum Error<T> {
+		/// Error when the number of orders for a trading pair has exceeded the
+		/// BoundedVec size for the order pair for the currency pair in
+		/// question.
+		AssetPairOrdersOverflow,
+		/// Error when order is placed attempting to exchange assets of the same
+		/// type.
+		ConflictingAssetIds,
+		/// Error when an account cannot reserve or transfer the amount
+		/// specified for trade, or amount to be fulfilled.
+		InsufficientAssetFunds,
+		/// Error when an account does not have enough funds in chains reserve
+		/// currency to place order.
+		InsufficientReserveFunds,
+		/// Error when account tries to buy an invalid amount of currency --
+		/// currently `0`.
+		InvalidBuyAmount,
+		/// Error when an account specifies an invalid buy price -- currently
+		/// `0`.
+		InvalidMinPrice,
+		/// Error when an order is placed with a currency that is not in the
+		/// asset registry.
+		InvalidAssetId,
+		/// Error when an operation is attempted on an order id that is not in
+		/// storage.
+		OrderNotFound,
+		/// Error when a user attempts an action on an order they are not
+		/// authorised to perform, such as cancelling another accounts order.
+		Unauthorised,
+		/// Error when unable to convert fee balance to asset balance when asset
+		/// out matches fee currency
+		BalanceConversionErr,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<T as frame_system::Config>::Hash: PartialEq<<T as frame_system::Config>::Hash>,
+	{
+		/// Create an order, with the minimum fulfillment amount set to the buy
+		/// amount, as the first iteration will not have partial fulfillment
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::Weights::create_order_v1())]
+		pub fn create_order_v1(
+			origin: OriginFor<T>,
+			asset_in: T::AssetCurrencyId,
+			asset_out: T::AssetCurrencyId,
+			buy_amount: T::ForeignCurrencyBalance,
+			price: T::ForeignCurrencyBalance,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			Self::place_order(
+				account_id, asset_in, asset_out, buy_amount, price, buy_amount,
+			)?;
+			Ok(())
+		}
+
+		///  Cancel an existing order that had been created by calling account.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::Weights::user_cancel_order())]
+		pub fn user_cancel_order(origin: OriginFor<T>, order_id: T::Hash) -> DispatchResult {
```

Comment:
Yeah, it's the name collision with the trait, so wanted to specify that this would be the one callable by users.

Reply:
I think it is totally fine having both names at the same time. If the user wants to call the trait once, they can use `<Self as Trait>::cancel_order()`. I also say that because `user_` is a generic word. It changes depending on the context: every call has a "user" who calls it.